$OpenBSD: patch-src_xvjpeg_c,v 1.1 2022/02/15 17:26:29 sthen Exp $

fix problems with Nikon D7100 pictures, they don't create correct exif
information (multiple APP1 chunks). xv reads them all, but libjpeg frowns
about writing multiple APP1.

So,
- ignore subsequent chunks (like exiv2 does)
- add a small framework to report error messages from libjpeg
(might be adaptable to other gfx formats).

Index: src/xvjpeg.c
--- src/xvjpeg.c.orig
+++ src/xvjpeg.c
@@ -87,9 +87,11 @@ static int   colorType;
 static DIAL  qDial, smDial;
 static BUTT  jbut[J_NBUTTS];
 
+char errbuffer[JMSG_LENGTH_MAX];
 
 
 
+
 /***************************************************************************/
 /* JPEG SAVE DIALOG ROUTINES ***********************************************/
 /***************************************************************************/
@@ -418,7 +420,7 @@ static void writeJPEG()
 
   if (pfree) free(inpix);
 
-  if (CloseOutFile(fp, filename, rv) == 0) DirBox(0);
+  if (CloseOutFileWhy(fp, filename, rv, errbuffer) == 0) DirBox(0);
   SetCursors(-1);
 }
 
@@ -444,7 +446,7 @@ METHODDEF void  xv_error_exit(cinfo)
   my_error_ptr myerr;
 
   myerr = (my_error_ptr) cinfo->err;
-  (*cinfo->err->output_message)(cinfo);     /* display error message */
+  (*cinfo->err->format_message)(cinfo, errbuffer);     /* fmt error message */
   longjmp(myerr->setjmp_buffer, 1);         /* return from error */
 }
 
@@ -851,7 +853,11 @@ METHODDEF boolean  xv_process_app1(cinfo)
     exifInfo = (byte *) malloc((size_t) length);
     exifInfoSize = 0;
   }
-  else exifInfo = (byte *) realloc(exifInfo, exifInfoSize + length);
+  else {
+    /* one APP1 data struct only, ignore extra stuff */
+    while (length-- > 0)
+      (void)j_getc(cinfo);
+  }
   if (!exifInfo) FatalError("out of memory in xv_process_app1 (EXIF info)");
   
   sp = exifInfo + exifInfoSize;
